File: app/javascript/controllers/entrega_personal_controller.js
import ClienteAutocomplete from "controllers/cliente_autocomplete"
import { conEnterAvanza } from "controllers/enter_avanza"

// PR-6: flow separado para entrega personal. Versión simplificada del
// etiquetar_controller — no necesita lookup de duplicado de tracking
// (el tracking se genera automático EP-YYYY-SUC-PROV-NNNNNN) ni
// detección de pre-alerta. Reutiliza el patrón de modal cantidad cajas.
export default class extends conEnterAvanza(ClienteAutocomplete) {
  static targets = [
    "form", "clienteInput", "clienteId", "clienteDropdown", "clienteNombre",
    "sucursalBanner", "sucursalTexto",
    "etiquetasModal", "etiquetasInput",
    "event", "panel"
  ]

  connect() {
    this._searchTimeout = null
    this._handleGlobalKeydown = this.handleKeydown.bind(this)
    document.addEventListener("keydown", this._handleGlobalKeydown)
  }

  disconnect() {
    document.removeEventListener("keydown", this._handleGlobalKeydown)
    if (this._searchTimeout) clearTimeout(this._searchTimeout)
  }

  handleKeydown(e) {
    if (e.key === "F2") {
      e.preventDefault()
      this.clearForm()
    } else if (e.key === "F9") {
      e.preventDefault()
      // C19-08, la gemela: mientras haya una pregunta abierta (la listita de
      // retener, la de política — cualquier <dialog>), F9 no guarda. Se
      // contesta con un Enter y se sigue; F2 queda libre.
      if (document.querySelector("dialog[open]")) return
      this.submitFormWithPrint()
    }
  }

  // Lo único que Entrega Personal hace de más al elegir un cliente: jalar sus
  // tareas y notas a la franja de la derecha (PR-9.b). El resto —búsqueda de
  // un dígito, preselección, flechas y Enter— es igual que en /etiquetar y
  // vive en `ClienteAutocomplete`.
  // C16-04: elegir el cliente con el teclado avanza de campo, igual que en
  // /etiquetar. El resto de la navegación por Enter viene del mixin.
  _despuesDeElegirConTeclado(e) {
    this._focusSiguiente(e.target)
  }

  _alSeleccionarCliente({ id, sucursalRetiro }) {
    // El mixin siempre mandó la sucursal de retiro; esta pantalla la tiraba, y
    // por eso nunca avisaba a dónde iba la caja. Yusef: "también misma
    // situación no avisa que va a tegus".
    this._mostrarSucursal(sucursalRetiro)
    this.loadPanel(id)
  }

  // Mismo aviso que /etiquetar, mismo motivo: decide en qué bolsa física cae la
  // caja, y enterarse tarde significa volver a abrirla (PR-C6.24).
  _mostrarSucursal(sucursal) {
    if (!this.hasSucursalBannerTarget) return

    const texto = (sucursal || "").trim()
    this.sucursalBannerTarget.classList.toggle("hidden", texto === "")
    if (texto !== "" && this.hasSucursalTextoTarget) {
      this.sucursalTextoTarget.textContent = texto
    }
  }

  // PR-C6.41: acá el tipo de envío es un select y puede cambiar después de
  // elegir al cliente, así que el panel de cálculo tiene que enterarse. En
  // /etiquetar no hace falta: el servicio es de la sesión y no se mueve.
  cambioTipoEnvio(e) {
    const calc = this.element.querySelector("[data-controller~='calc-volumetrico']")
    if (!calc) return

    calc.dataset.calcVolumetricoTipoEnvioIdValue = parseInt(e.target.value, 10) || 0
  }

  // Recarga el turbo-frame de la franja. Sin tracking que pasar: en entrega
  // personal el tracking lo genera el sistema al guardar (EP-AÑO-SUC-PROV-N).
  loadPanel(clienteId) {
    if (!this.hasPanelTarget) return

    const frame = this.panelTarget.querySelector("turbo-frame#panel_contexto")
    if (!frame) return

    const base = this.panelTarget.dataset.panelUrl
    const url = clienteId ? `${base}?cliente_id=${encodeURIComponent(clienteId)}` : base
    if (frame.getAttribute("src") !== url) frame.setAttribute("src", url)
  }


  // Submit + modal de cajas (mismo patrón de PR-4 etiquetar).
// PR-C6.31 sacó de acá un modal "¿cuántas cajas?" que preguntaba lo MISMO que
// el campo visible del formulario —y peor: lo reseteaba a 1 antes de
// preguntar—. Ese modal estaba mal, y sigue estándolo.
//
// C20-07 trae el de /etiquetar (PR-C7.23), que es otra cosa: solo aparece
// cuando NO se midió ninguna caja, así que nunca compite con las filas. Y acá
// hacía más falta que allá, porque en Entrega Personal casi nunca se mide: sin
// esto, un envío de tres bultos no tenía forma de pedir tres etiquetas. Jorge:
// "esta lógica de las cajas que hicimos para etiquetar también hay que
// aplicarla en entrega personal".
submitFormWithPrint() {
  if (this._cajasCargadas() > 0) return this._submitWithPrint()
  if (!this.hasEtiquetasModalTarget) return this._submitWithPrint()

  if (this.hasEtiquetasInputTarget) this.etiquetasInputTarget.value = "1"
  this.dispatch("modalAbierto")
  this.etiquetasModalTarget.showModal()
  // `select()` y no solo `focus()`: el operario teclea el número encima sin
  // tener que borrar el 1.
  if (this.hasEtiquetasInputTarget) this.etiquetasInputTarget.select()
}

  // Cuántas filas de caja hay cargadas. Se cuentan las filas —la misma fuente
  // que usa el repetidor—, más la que se esté tecleando arriba sin haberle
  // dado «Agregar», que es lo que la pantalla misma dice que se puede hacer
  // (C18-05).
  _cajasCargadas() {
    const filas = this.formTarget.querySelectorAll(".caja-fila").length
    const captura = this.formTarget.querySelector("[data-caja-campo='peso']")
    return filas + (captura && captura.value.trim() !== "" ? 1 : 0)
  }

  // Enter confirma; Escape cancela. Acá Enter SÍ actúa, al revés que en el
  // formulario —donde la pistola dispara Enter y por eso pasa al campo
  // siguiente—: el modal no es el formulario y el operario ya decidió imprimir.
  etiquetasKeydown(e) {
    if (e.key === "Enter") {
      e.preventDefault()
      this.confirmarEtiquetas()
    } else if (e.key === "Escape") {
      e.preventDefault()
      this.cerrarEtiquetas()
    }
  }

  confirmarEtiquetas() {
    const cantidad = this._etiquetasPedidas()
    if (cantidad === null) return

    this.etiquetasModalTarget.close()
    this._submitWithPrint(cantidad)
  }

  cerrarEtiquetas() {
    this.etiquetasModalTarget.close()
  }

  // Un número mal tecleado no puede grabar 500 paquetes ni tirar 500
  // etiquetas: el server tiene el mismo tope, esto es para avisar antes.
  _etiquetasPedidas() {
    const n = parseInt(this.etiquetasInputTarget.value, 10)
    if (!Number.isInteger(n) || n < 1 || n > 99) {
      this.etiquetasInputTarget.select()
      return null
    }
    return n
  }

  _submitWithPrint(etiquetas = null) {
    this._removePrintField()
    const input = document.createElement("input")
    input.type = "hidden"
    input.name = "print"
    input.value = "true"
    input.dataset.printField = "true"
    this.formTarget.appendChild(input)

    // Suelto y NO como `paquete[cantidad_paquetes]`: el bug de PR-C6.31 fue
    // tener dos campos con el mismo `name`.
    if (etiquetas) {
      const cuantas = document.createElement("input")
      cuantas.type = "hidden"
      cuantas.name = "etiquetas"
      cuantas.value = String(etiquetas)
      cuantas.dataset.printField = "true"
      this.formTarget.appendChild(cuantas)
    }

    this.formTarget.requestSubmit()
  }

  // `querySelectorAll` y no `querySelector`: desde que el modal agrega también
  // la cantidad son dos. Con el singular, el segundo sobrevivía al guardado y
  // el paquete siguiente heredaba la cantidad del anterior — el bug de
  // PR-C6.31, que en /etiquetar ya se pagó una vez.
  _removePrintField() {
    this.formTarget.querySelectorAll("[data-print-field]").forEach(el => el.remove())
  }

  clearForm() {
    this.formTarget.reset()
    // Las cajas cargadas son de ESTE paquete: se van con él. `reset()` no las
    // toca porque viven en inputs hidden que el repetidor arma a mano.
    this.formTarget.dispatchEvent(new CustomEvent("cajas:limpiar", { bubbles: true }))
    if (this.hasClienteIdTarget) this.clienteIdTarget.value = ""
    if (this.hasClienteNombreTarget) {
      this.clienteNombreTarget.textContent = ""
      this.clienteNombreTarget.classList.add("hidden")
    }
    // PR-9.b: la franja vuelve a su estado vacío junto con el formulario.
    this.loadPanel(null)

    // C19-02: acá no se enfocaba nada y el cursor quedaba "como en el aire".
    // Vuelve al mismo campo que arranca la pantalla (el [autofocus], hoy el
    // proveedor), para encadenar el siguiente paquete sin agarrar el mouse.
    const primero = this.formTarget.querySelector("[autofocus]")
    if (primero) primero.focus()
  }

  // Handle turbo stream events después del save.
  eventTargetConnected(el) {
    if (el.dataset.action !== "paquete-saved") return
    this.dispatch("success")
    if (el.dataset.print === "true") {
      // PR-10.d: la ETIQUETA (Dymo 2.25x1.25), no el Warehouse Receipt.
      // Yusef: "aqui esta tirando el warehouse, no la etiqueta".
      // `hermanas=1` saca una por caja cuando el tracking se dividio.
      //
      // PR-C7.16: el controller marca `data-print` solo en la primera caja, asi
      // que esto corre una vez por tracking y no una por caja.
      //
      // Y despues el Warehouse Receipt, que es uno solo para todo el envio.
      // Yusef: "la etiqueta es la que le pegamos a cada caja, pero pido tres;
      // el warehouse receipt es al reves: solo imprimis uno, donde detalla todo
      // lo que recibiste". Iba como preview y lo cambió probándolo (C19-01):
      // ahora imprime de un solo y la pestaña se cierra, como la etiqueta; el
      // preview queda en el botón «Ver WR» de la ficha y el link del flash.
      //
      // `wr=1` y **una sola ventana**: esto eran dos `window.open` seguidas y
      // Chrome bloqueaba la segunda —un gesto del usuario da permiso para un
      // popup, no para dos—. Yusef veia salir la etiqueta y nada mas. Ahora la
      // ventana de la etiqueta, al terminar de imprimir, se va al Warehouse
      // Receipt en vez de cerrarse (ver `layouts/etiqueta.html.erb`).
      window.open(`/paquetes/${el.dataset.paqueteId}/etiqueta?hermanas=1&print=true&wr=1`, "_blank")
      // C19-02: la pestaña de impresión se lleva el foco de la ventana; cuando
      // el WR se cierra y la ventana vuelve, de regreso al primer campo — el
      // mismo viaje que hace /etiquetar con el tracking. El guard de
      // isConnected es por si Turbo reemplazó la página en el medio.
      window.addEventListener("focus", () => {
        if (!this.element.isConnected) return
        const primero = this.formTarget.querySelector("[autofocus]")
        if (primero) primero.focus()
      }, { once: true })
    }
    setTimeout(() => this.clearForm(), 100)
    el.remove()
  }
}

File: app/controllers/entrega_personal_controller.rb
class EntregaPersonalController < ApplicationController
  # PR-C6.31: el partial de peso y medidas es el mismo que el de /etiquetar,
  # así que las filas por caja también tienen que leerse igual. Antes esta
  # pantalla las pintaba y después las tiraba: todas las cajas nacían con el
  # peso de arriba.
  include MedidasPorCaja
  # El sellado del prepago vive en el concern porque /etiquetar hace lo mismo.
  include PrepagoMiami
  include NotificaRecibido

  before_action :authorize_entrega_personal

  # PR-6: flow separado para paquetes que llegan FÍSICAMENTE al mostrador
  # en Miami (sin tracking del courier). Sistema genera tracking
  # automático EP-YYYY-SUC-PROV-NNNNNN vía Paquete#generate_ep_tracking
  # callback que ya existe (PR-D3.b).

  def new
    @sucursales_recepcion = sucursales_de_recepcion_con_ep
    # La misma preselección que /etiquetar (gemelas): la sucursal del usuario,
    # si no la de recepción por defecto. Antes el select nacía en blanco.
    @paquete = Paquete.new(sucursal_recepcion: Sucursal.recepcion_por_defecto_para(Current.user, entre: @sucursales_recepcion))
    @paquetes_hoy = paquetes_ep_hoy_count
    @tipo_envios = TipoEnvio.activos.order(:nombre)
    @proveedores_ep = Proveedor.where(tipo: "entrega_personal").activos.ordered
    @motivos_retencion = MotivoRetencion.activos.ordered
    @motivos_envio_politica = MotivoEnvioPolitica.activos.ordered
    @tarifas_recolecta = TarifaRecolecta.activas.ordered
  end

  # A7-20. La cantidad de cajas **se deriva de las filas**, no de un campo.
  #
  # Antes venía de un `cantidad_paquetes` que el operario tecleaba, y ese
  # desacople ya había costado un bug: en PR-C6.31 el form mandaba el campo dos
  # veces, ganaba el hidden con valor 1, y el operario veía tres filas pero se
  # grababa un solo paquete. Si el número sale de contar las filas, no hay dos
  # fuentes que puedan discrepar.
  def create
    cajas = medidas_por_caja

    case cajas.size
    when 0 then create_single                          # nunca tocó Agregar: un solo bulto
    when 1 then create_single(cajas.values.first)      # una caja agregada: sus datos mandan
    else        create_split(cajas.size)
    end
  end

  private

  def create_single(medidas = {})
    @paquete = Paquete.new(paquete_params.merge(medidas))
    apply_extra_params(@paquete)
    heredar_sucursal_de_retiro(@paquete)
    @paquete.estado = "recibido_miami"
    @paquete.user = Current.user

    if @paquete.save
      respond_saved([ @paquete ])
    else
      render_create_error
    end
  end

  def create_split(total_cajas)
    attrs = con_sucursal_de_retiro(
      paquete_params.except(:cantidad_paquetes, :numero_caja)
    ).merge(
      estado: "recibido_miami",
      user: Current.user
    )
    paquetes = Paquete.crear_split!(attrs: attrs, total_cajas: total_cajas,
                                    por_caja: medidas_por_caja)
    paquetes.each do |p|
      apply_extra_params(p)
      heredar_sucursal_de_retiro(p)
      p.save!
    end
    respond_saved(paquetes)
  rescue ActiveRecord::RecordInvalid => e
    @paquete = e.record
    render_create_error
  end

  # PR-C6.37: donde retira el cliente. Ojo con el choque de nombres: en esta
  # pantalla `sucursal_id` es la de MIAMI donde se recibe (define el prefijo del
  # tracking EP), asi que la de retiro solo se pone si el form no mando una.
  def heredar_sucursal_de_retiro(paquete)
    paquete.sucursal ||= paquete.cliente&.sucursal_retiro
  end

  # `crear_split!` crea las cajas de una, así que la herencia de la sucursal de
  # retiro tiene que estar puesta **antes** — si no, la primera caja se valida
  # con el campo vacío. Por eso viaja en `attrs` y no solo en el `each` de
  # después.
  def con_sucursal_de_retiro(attrs)
    return attrs if attrs[:sucursal_id].present?

    retiro = Cliente.find_by(id: attrs[:cliente_id])&.sucursal_retiro
    retiro ? attrs.merge(sucursal_id: retiro.id) : attrs
  end

  # `prepagado_miami` se traduce a cinco columnas, así que no puede ir en
  # `paquete_params`. El sellado lo hace `PrepagoMiami`, compartido con
  # /etiquetar: escrito acá otra vez, las dos pantallas se separan.
  def apply_extra_params(paquete, save: false)
    if (prov_str = proveedor_string_param) != :missing
      paquete[:proveedor] = prov_str
    end
    aplicar_prepago_miami(paquete)
    paquete.save! if save
  end

  # C17-02: acá **no** se llama a `Tarea.atar_al_paquete!`, a propósito. El
  # tracking de una entrega personal se genera al guardar (EP-AÑO-SUC-PROV-N),
  # así que la tarea que se deja desde la franja no tiene tracking y no hay por
  # dónde atarla; «la última que dejó este usuario» falla apenas se intercalan
  # dos tareas o dos paquetes. Queda del cliente y sale en la bandeja.
  def respond_saved(paquetes)
    @paquete = paquetes.first
    # C18-06: en Entrega Personal no hay pre-alerta; el correo sale solo si se
    # marcó «enviado según política».
    notificar_recibido(@paquete, pre_alerta_vinculada: false)
    msg = paquetes.size > 1 ?
            "Entrega personal registrada — #{paquetes.size} cajas, tracking #{@paquete.tracking}" :
            "Entrega personal registrada — tracking #{@paquete.tracking}"

    respond_to do |format|
      format.turbo_stream do
        # PR-C7.16: **solo la primera caja dispara la impresión.** Las N cajas de
        # un split comparten tracking desde `crear_split!`, así que
        # `etiqueta?hermanas=1` ya saca las N etiquetas en un solo trabajo.
        # Marcar las tres abría tres ventanas imprimiendo lo mismo tres veces.
        #
        # Yusef: "aquí debería crear las etiquetas para 3 cajas y luego tirar
        # preview del WR" — el WR es uno solo, "al contrario" de la etiqueta.
        events = paquetes.each_with_index.map do |p, i|
          imprime = i.zero? && params[:print].to_s == "true"
          "<div data-entrega-personal-target='event' data-action='paquete-saved' " \
            "data-guia='#{p.guia}' data-print='#{imprime}' data-paquete-id='#{p.id}'></div>"
        end.join
        render turbo_stream: [
          turbo_stream.update("paquetes-counter", paquetes_ep_hoy_count.to_s),
          turbo_stream.prepend("flash-messages", partial: "shared/flash",
                                                 locals: { notice: aviso_con_wr(msg) }),
          turbo_stream.append("entrega-personal-events", events)
        ]
      end
      format.html do
        redirect_to new_entrega_personal_path, notice: msg
      end
    end
  end

  # El aviso lleva el link al Warehouse Receipt.
  #
  # Yusef: *"aquí debería crear las etiquetas para 3 cajas y **luego tirar
  # preview del WR**"*. El JS lo intenta abrir en una ventana, pero Chrome
  # permite **un solo popup por gesto del usuario** y ese ya se lo lleva la
  # ventana de impresión de las etiquetas — que es la que no puede faltar.
  #
  # Así que el WR va también acá, donde no depende del permiso de popups. El
  # texto se escapa; solo el link es HTML.
  def aviso_con_wr(msg)
    return msg if @paquete.nil?

    helpers.safe_join([
      msg,
      " · ",
      helpers.link_to("Ver Warehouse Receipt",
                      warehouse_receipt_paquete_path(@paquete),
                      target: "_blank", rel: "noopener",
                      class: "underline font-medium")
    ])
  end

  # C18-02: las mismas que ofrece /etiquetar (`Sucursal.de_recepcion`), y de
  # ellas las que tienen código EP —sin él no se genera el tracking
  # EP-AÑO-SUC-…—. Acá decía `where(ubicacion: "miami")` a mano: la gemela
  # separada, y México (`otros`) tampoco habría salido.
  def sucursales_de_recepcion_con_ep
    Sucursal.de_recepcion.con_codigo_ep
  end

  def render_create_error
    @paquetes_hoy = paquetes_ep_hoy_count
    @tipo_envios = TipoEnvio.activos.order(:nombre)
    @proveedores_ep = Proveedor.where(tipo: "entrega_personal").activos.ordered
    # PR-9.b: faltaba esto — la vista hace `.any?` sobre él, así que cualquier
    # error de validación reventaba con un 500 en vez de mostrar los errores.
    @sucursales_recepcion = sucursales_de_recepcion_con_ep
    @motivos_retencion = MotivoRetencion.activos.ordered
    @motivos_envio_politica = MotivoEnvioPolitica.activos.ordered
    @tarifas_recolecta = TarifaRecolecta.activas.ordered
    # Igual que en /etiquetar: las cajas medidas vuelven a la pantalla. Sin
    # esto, un error de validación las borraba todas.
    @cajas_cargadas = medidas_por_caja
    flash.now[:alert] = "No se pudo registrar la entrega personal."
    render :new, status: :unprocessable_entity
  end

  def authorize_entrega_personal
    require_role(:supervisor_miami, :digitador_miami)
  end

  def paquetes_ep_hoy_count
    Paquete
      .joins(:proveedor)
      .where(proveedores: { tipo: "entrega_personal" })
      .where(user: Current.user)
      .where(created_at: Time.current.beginning_of_day..Time.current.end_of_day)
      .count
  end

  def paquete_params
    params.require(:paquete).permit(
      # PR-C7.16: la sucursal de Miami entra como `sucursal_recepcion_id`.
      # `sucursal_id` sigue permitido porque es **dónde retira el cliente**, y el
      # form podría llegar a ofrecerlo; hoy lo hereda `heredar_sucursal_de_retiro`.
      :cliente_id, :tipo_envio_id, :proveedor_id, :sucursal_id, :sucursal_recepcion_id, :peso,
      :alto, :largo, :ancho, :cantidad_productos, :cantidad_paquetes,
      :numero_caja, :descripcion, :remitente, :driver,
      :notas_internas, :notas_retencion,
      :retener_miami, :enviado_por_politica, :notas_envio_politica,
      # A7-22/A7-23: la recolecta vive en esta misma pantalla.
      :recolecta_solicitada, :tarifa_recolecta_id, :recolecta_monto, :recolecta_moneda,
      :recolecta_contacto, :recolecta_telefono, :recolecta_horario, :recolecta_instrucciones,
      :recolecta_direccion,
      motivo_retencion_ids: [], motivo_envio_politica_ids: []
    )
  end

  # `proveedor` string legacy choca con `belongs_to :proveedor` (mismo
  # patrón del bug PR fix #185 en EtiquetarController). Lo extraemos
  # aparte para asignar vía column accessor.
  def proveedor_string_param
    return :missing unless params.dig(:paquete)&.key?(:proveedor)

    params[:paquete][:proveedor].to_s
  end
end
